ghcide/src/Development/IDE/Plugin/CodeAction.hs

```diff
@@ -224,6 +226,69 @@ caRemoveRedundantImports m contents digs ctxDigs uri
         _edit = Just WorkspaceEdit{..}
         _command = Nothing
 
+caRemoveRedundantExports :: Maybe ParsedModule -> Maybe T.Text -> [Diagnostic] -> [Diagnostic] -> Uri -> [CAResult]
+caRemoveRedundantExports m contents digs ctxDigs uri
+  | Just pm <- m,
+    Just txt <- contents,
+    txt' <- indexedByPosition $ T.unpack txt,
+    r <- mapMaybe (groupDiag pm) digs,
+    r' <- map (\(t,d,rs) -> (t,d,extend txt' rs)) r,
+    caRemoveCtx <- mapMaybe removeSingle r',
+    allRanges <- nub $ [ range | (_,_,ranges) <- r, range <- ranges],
+    allRanges' <- extend txt' allRanges,
+    Just caRemoveAll <- removeAll allRanges',
+    ctxEdits <- [ x | x@(_, d, _) <- r, d `elem` ctxDigs],
+    not $ null ctxEdits
+      = caRemoveCtx ++ [caRemoveAll]
+  | otherwise = []
+  where
+    extend txt ranges = extendAllToIncludeCommaIfPossible True txt ranges
```

Comment:
These plural code actions do not scale well, since they process all the D diagnostics on every call. So the complexity is `D^2`. 

And in fact, they are unnecessary. VSCode supports applying all the quick fixes on save, just by setting the config below. It's even possible for LSP servers to mark some code actions as not-preferred, to prevent them from being applied with this setting, although this is currently not exposed by `haskell-lsp-types` and I'm unsure how it works. Other editors probably have similar options.
 
In `settings.json`:
```
  "editor.codeActionsOnSave": {
    "source.fixAll": true
  }
```

Reply:
> These plural code actions do not scale well, since they process all the D diagnostics on every call. So the complexity is D^2.

I don't understand which D^2 you mean. `caRemoveRedundantExports` is called once per `codeAction`. `nub` is O^2 (on some diagnostics) but it can be removed with a bit more careful design.

Also, I'm not sure how editors applies all changes together. These code actions remove parts of the code, however some parts may interleave.

Reply:
haskell-lsp calls the code action handler once per diagnostic, so D times. The handler calls `caRemoveRedundantExports` which does D work each time, so D^2 work in total.

There are ways around it, like postponing the work until the code action is actually executed, but given that clients can just take all the code actions and apply them at once, it doesn't seem worth the effort.